lesson08: exercise and assignment

bromleysj opened the pull request:
**Note: circle.py and test_circle.py are complete, however a few tests fail: test_iadd and test_imult**
The errors I get are: 
AttributeError: 'NoneType' object has no attribute 'radius'
AttributeError: 'float' object has no attribute 'radius'
I'd appreciate some direction so I can push a fix.

natasha-aleksandrova commented on students/scott_bromley/lesson08/circle.py:
```diff
@@ -0,0 +1,179 @@
+        if isinstance(self, Circle) & isinstance(other, int):
+            return self.__add__(other.radius)
+        if isinstance(self, Circle) & isinstance(other, Circle):
+            return self.__add__(other.radius)
```
general comment: you should use `and` for comparison

not sure why the if checks when you always return the same thing, but this method should look like this:
```
self.radius += other.radius
return self
```

natasha-aleksandrova commented on students/scott_bromley/lesson08/test_circle.py:
```diff
@@ -0,0 +1,351 @@
+
+    @mark.add
+    @mark.parametrize("test_input, expected", [
+        ([Circle(10), 2.25], 12.25),
```
you're passing in an integer but your code expects a Circle instance?

natasha-aleksandrova commented on students/scott_bromley/lesson08/circle.py:
```diff
@@ -0,0 +1,179 @@
+        if isinstance(self, Circle) & isinstance(other, int):
+            return self.__mul__(other.radius)
+        if isinstance(self, Circle) & isinstance(other, Circle):
+            return self.__mul__(other.radius)
```
same comment as in iadd

natasha-aleksandrova commented on students/scott_bromley/lesson08/circle.py:
```diff
@@ -0,0 +1,179 @@
+            return False
+
+    def __gt__(self, other):
+        if self.radius > other.radius:
```
this can be simplified to just `return self.radius > other.radius`

natasha-aleksandrova commented:
merging, feel free to resubmit when ready.

bromleysj commented:
Natasha,

I resubmitted the assignments for weeks 7 and 8.

I have a question about the week 9 assignment: I'm having troubles with the
DonorCollection class.

My __init__ currently is:

def __init__(self, *donors):

My problem here is that I know I'm going to be passing in a list of Donor
objects, but is treating this as a list considered good coding practice?

Also, I know I'll need to create a method to "add" a Donor object, is the
goal here to use something like def append(self, donor) in the spirit of a
collection re: list?

Also, I've provided sort functionality to the DonorCollection class, am I
correct that this is where that sort functionality belongs?

Thanks,

Scott




On Tue, Mar 3, 2020 at 7:09 PM Natasha <notifications@github.com> wrote:

> Merged #758
> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758>
> into master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758?email_source=notifications&email_token=ACI2LMK73UUMCOXW23PFUG3RFWL4TA5CNFSM4K7ARUAKYY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOXBYKSFI#event-3094391061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACI2LMJB43I75VWEJF6Q57DRFWL4TANCNFSM4K7ARUAA>
> .
>

bromleysj commented:
Natasha,

On second thought, I'm wondering if it isn't better for the internal
representation in DonorCollection to be a defaultdict?

I apologize for the many e-mails.

-Scott

On Sun, Mar 8, 2020 at 8:06 PM Scott Bromley <bromleysj@gmail.com> wrote:

> Natasha,
>
> I resubmitted the assignments for weeks 7 and 8.
>
> I have a question about the week 9 assignment: I'm having troubles with
> the DonorCollection class.
>
> My __init__ currently is:
>
> def __init__(self, *donors):
>
> My problem here is that I know I'm going to be passing in a list of Donor
> objects, but is treating this as a list considered good coding practice?
>
> Also, I know I'll need to create a method to "add" a Donor object, is the
> goal here to use something like def append(self, donor) in the spirit of a
> collection re: list?
>
> Also, I've provided sort functionality to the DonorCollection class, am I
> correct that this is where that sort functionality belongs?
>
> Thanks,
>
> Scott
>
>
>
>
> On Tue, Mar 3, 2020 at 7:09 PM Natasha <notifications@github.com> wrote:
>
>> Merged #758
>> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758>
>> into master.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758?email_source=notifications&email_token=ACI2LMK73UUMCOXW23PFUG3RFWL4TA5CNFSM4K7ARUAKYY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOXBYKSFI#event-3094391061>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACI2LMJB43I75VWEJF6Q57DRFWL4TANCNFSM4K7ARUAA>
>> .
>>
>

natasha-aleksandrova commented:
@bromleysj yes dictionary would be much better here because you are still looking up by Donor Names.

So you could have a dictionary with a key being a name (string) and a value being Donor object.

Hope this helps.

bromleysj commented:
Natasha,

I know I'm making some stupid, obvious error here but it just isn't jumping
out to me.
I'm constantly getting back an error of 'Donor' object has no attribute
'_donations'.
In my __init__ I want to allow for the user to create a Donor object
without passing in any donations: is it better to do this through
a @classmethod such as def from_donor_name or to allow it in the
constructor as I'm trying below?
Thanks so much,
Scott



class Donor:
    """
    Donor class represents a single donor
    """
    def __init__(self, donor_name="", donations=None):
        """
        Donor object initialization
        :param donor_name: full name of donor
        :param donations: donations
        """
        self._donor_name = donor_name
        if hasattr(donations, "__iter__"):
            try:
                [self.add_donation(donation) for donation in donations]
            except TypeError:
                print(f"{donations} is not iterable")
        else:
            self.add_donation(donations)

    def __call__(self, donor_name, donations):
        return self._donor_name, self.donations

    @staticmethod
    def from_donor_name(cls, donor_name):
        """
        alternate constructor w/ no donations
        :param cls: class object
        :param donor_name: donor name
        :return:
        """
        return cls(donor_name)

    @property
    def donor_name(self):
        """
        getter for donor_name
        :return: name of donor
        """
        return self._donor_name

    @donor_name.setter
    def donor_name(self, name):
        """
        setter for donor_name
        :param name: name of donor
        :return: None
        """
        if not isinstance(name, str):
            raise TypeError("donor name must be a string")
        if not name:
            raise ValueError("invalid donor name")
        self._donor_name = name
        #name_regexp = "[A-Za-z]{2,25}\s[A-Za-z]{2,25}"
        # try:
        #     self.donor_name = ''.join(re.findall(name_regexp,
name)).title().strip()
        # except ValueError as name_err:
        #     raise name_err("invalid donor name")

    @property
    def donations(self):
        """
        getter for donations
        :return: list of donations
        """
        return self._donations

    @property
    def first_name(self):
        """
        getter for donor first name
        :return: first name of donor
        """
        return self.donor_name.split()[0]

    @property
    def last_name(self):
        """
        getter for donor last name
        :return: last name of donor
        """
        return self.donor_name.split()[1]

    @property
    def num_donations(self):
        """
        getter for number of donations
        :return: number of donations
        """
        return len(self.donations)

    @property
    def total_donations(self):
        """
        getter for sum of donations
        :return: donations sum for donor
        """
        return sum(self.donations)

    @property
    def avg_donation(self):
        """
        getter for average donation
        :return: average donation of donor
        """
        if self.num_donations == 0:
            raise ValueError(f"{self.donor_name} has no donations")
        return round(mean(self.donations), 2)

    def add_donation(self, donation: float):
        """
        adds donation to Donor object
        :param self:
        :param donation: donation amount
        :return: None
        """
        if not donation or donation < 0:
            raise ValueError("invalid donation amount")
        try:
            self.donations.append(float(donation))
        except TypeError as type_err:
            raise type_err("invalid donation type")

    def thank_you(self):
        """
        create formatted letter string for donor
        :return: formatted letter
        """
        return f"\n\n{datetime.now.date()},\n\n\n" \
               f"Dear {self.donor_name},\n\n\n" \
               f"On behalf of all of us we thank you for your generous\n" \
               f"donation of ${self.donations[-1]:03.2f}.\n" \
               f"To date, you have donated: ${self.total_donations:03.2f}.\n" \
               "We look forward to seeing you at our annual meeting.\n\n" \
               "Sincerely,\n\n" \
               "ADL\n"

    def report_row(self):
        """
        format report row for Donor object
        :return: formatted report row Donor
        """
        return ('{:<25}{:^5}${:>14,.2f}{:^5}{:>10}{:^5}${:>14,.2f}'
                ).format(self.donor_name, ' ',
                         self.total_donations, ' ',
                         self.num_donations, ' ',
                         self.avg_donation
                         )


On Mon, Mar 9, 2020 at 11:00 PM Natasha <notifications@github.com> wrote:

> @bromleysj <https://github.com/bromleysj> yes dictionary would be much
> better here because you are still looking up by Donor Names.
>
> So you could have a dictionary with a key being a name (string) and a
> value being Donor object.
>
> Hope this helps.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758?email_source=notifications&email_token=ACI2LMKZCZI5L57BQFNKSNTRGW3NPA5CNFSM4K7ARUAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOJZ3RI#issuecomment-596876741>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACI2LMMNL6UALQTHNPU443LRGW3NPANCNFSM4K7ARUAA>
> .
>

bromleysj commented:
Also, I'm worried about hitting the deadline. Is there any chance it can be
extended? I think I can still make the 18th, just worried.

On Thu, Mar 12, 2020, 10:57 PM Scott Bromley <bromleysj@gmail.com> wrote:

> Natasha,
>
> I know I'm making some stupid, obvious error here but it just isn't
> jumping out to me.
> I'm constantly getting back an error of 'Donor' object has no attribute
> '_donations'.
> In my __init__ I want to allow for the user to create a Donor object
> without passing in any donations: is it better to do this through
> a @classmethod such as def from_donor_name or to allow it in the
> constructor as I'm trying below?
> Thanks so much,
> Scott
>
>
>
> class Donor:
>     """
>     Donor class represents a single donor
>     """
>     def __init__(self, donor_name="", donations=None):
>         """
>         Donor object initialization
>         :param donor_name: full name of donor
>         :param donations: donations
>         """
>         self._donor_name = donor_name
>         if hasattr(donations, "__iter__"):
>             try:
>                 [self.add_donation(donation) for donation in donations]
>             except TypeError:
>                 print(f"{donations} is not iterable")
>         else:
>             self.add_donation(donations)
>
>     def __call__(self, donor_name, donations):
>         return self._donor_name, self.donations
>
>     @staticmethod
>     def from_donor_name(cls, donor_name):
>         """
>         alternate constructor w/ no donations
>         :param cls: class object
>         :param donor_name: donor name
>         :return:
>         """
>         return cls(donor_name)
>
>     @property
>     def donor_name(self):
>         """
>         getter for donor_name
>         :return: name of donor
>         """
>         return self._donor_name
>
>     @donor_name.setter
>     def donor_name(self, name):
>         """
>         setter for donor_name
>         :param name: name of donor
>         :return: None
>         """
>         if not isinstance(name, str):
>             raise TypeError("donor name must be a string")
>         if not name:
>             raise ValueError("invalid donor name")
>         self._donor_name = name
>         #name_regexp = "[A-Za-z]{2,25}\s[A-Za-z]{2,25}"
>         # try:
>         #     self.donor_name = ''.join(re.findall(name_regexp, name)).title().strip()
>         # except ValueError as name_err:
>         #     raise name_err("invalid donor name")
>
>     @property
>     def donations(self):
>         """
>         getter for donations
>         :return: list of donations
>         """
>         return self._donations
>
>     @property
>     def first_name(self):
>         """
>         getter for donor first name
>         :return: first name of donor
>         """
>         return self.donor_name.split()[0]
>
>     @property
>     def last_name(self):
>         """
>         getter for donor last name
>         :return: last name of donor
>         """
>         return self.donor_name.split()[1]
>
>     @property
>     def num_donations(self):
>         """
>         getter for number of donations
>         :return: number of donations
>         """
>         return len(self.donations)
>
>     @property
>     def total_donations(self):
>         """
>         getter for sum of donations
>         :return: donations sum for donor
>         """
>         return sum(self.donations)
>
>     @property
>     def avg_donation(self):
>         """
>         getter for average donation
>         :return: average donation of donor
>         """
>         if self.num_donations == 0:
>             raise ValueError(f"{self.donor_name} has no donations")
>         return round(mean(self.donations), 2)
>
>     def add_donation(self, donation: float):
>         """
>         adds donation to Donor object
>         :param self:
>         :param donation: donation amount
>         :return: None
>         """
>         if not donation or donation < 0:
>             raise ValueError("invalid donation amount")
>         try:
>             self.donations.append(float(donation))
>         except TypeError as type_err:
>             raise type_err("invalid donation type")
>
>     def thank_you(self):
>         """
>         create formatted letter string for donor
>         :return: formatted letter
>         """
>         return f"\n\n{datetime.now.date()},\n\n\n" \
>                f"Dear {self.donor_name},\n\n\n" \
>                f"On behalf of all of us we thank you for your generous\n" \
>                f"donation of ${self.donations[-1]:03.2f}.\n" \
>                f"To date, you have donated: ${self.total_donations:03.2f}.\n" \
>                "We look forward to seeing you at our annual meeting.\n\n" \
>                "Sincerely,\n\n" \
>                "ADL\n"
>
>     def report_row(self):
>         """
>         format report row for Donor object
>         :return: formatted report row Donor
>         """
>         return ('{:<25}{:^5}${:>14,.2f}{:^5}{:>10}{:^5}${:>14,.2f}'
>                 ).format(self.donor_name, ' ',
>                          self.total_donations, ' ',
>                          self.num_donations, ' ',
>                          self.avg_donation
>                          )
>
>
> On Mon, Mar 9, 2020 at 11:00 PM Natasha <notifications@github.com> wrote:
>
>> @bromleysj <https://github.com/bromleysj> yes dictionary would be much
>> better here because you are still looking up by Donor Names.
>>
>> So you could have a dictionary with a key being a name (string) and a
>> value being Donor object.
>>
>> Hope this helps.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758?email_source=notifications&email_token=ACI2LMKZCZI5L57BQFNKSNTRGW3NPA5CNFSM4K7ARUAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOJZ3RI#issuecomment-596876741>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACI2LMMNL6UALQTHNPU443LRGW3NPANCNFSM4K7ARUAA>
>> .
>>
>

natasha-aleksandrova commented:
@bromleysj you never define `_donations` but you use it in a property.

if you would like an optional donation for your class then use optional keyword `donations=None` in the constructor and initialize it an empty list.

natasha-aleksandrova commented:
@bromleysj for an extension... please send us an email with a request 🙂

bromleysj commented:
Natasha,

I respectfully request an extension.  My son's daycare has been closed due
to COVID-19 (coronavirus) concerns and I've had little time to dedicate to
coursework.

Please let me know if this is approved, I'm still going to try to finish by
the 18th but between work and having to watch my son during the day it will
be tough.

Thanks,

Scott

On Fri, Mar 13, 2020 at 12:42 PM Natasha <notifications@github.com> wrote:

> @bromleysj <https://github.com/bromleysj> for an extension... please send
> us an email with a request 🙂
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UWPCE-PythonCert-ClassRepos/SP_Online_PY210/pull/758#issuecomment-598813701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACI2LMJBKY6A3XP74RZIKRLRHJPBHANCNFSM4K7ARUAA>
> .
>

natasha-aleksandrova commented:
@bromleysj please send us an email.

(this thread is publicly visible on github)
